ark/type/keywords/string/base64.ts

```diff
@@ -1,24 +1,125 @@
+const revLookup = lookup.reduce(
+	(obj, char, i) => {
+		obj[char.charCodeAt(0)] = i
+		return obj
+	},
+	{} as Record<number, number>
+)
+
+const getLens = (b64: string) => {
+	const len = b64.length
+
+	if (len % 4 > 0)
+		throw new SyntaxError("Invalid string. Length must be a multiple of 4")
+
+	// Trim off extra bytes after placeholder bytes are found
+	// See: https://github.com/beatgammit/base64-js/issues/42
+	let validLen = b64.indexOf("=")
+	if (validLen === -1) validLen = len
+
+	const placeHoldersLen = validLen === len ? 0 : 4 - (validLen % 4)
+
+	return [validLen, placeHoldersLen]
+}
+
+const byteLength = (validLen: number, placeHoldersLen: number) =>
+	((validLen + placeHoldersLen) * 3) / 4 - placeHoldersLen
+
+const parseB64 = (b64: string) => {
+	const [validLen, placeHoldersLen] = getLens(b64)
+	const arr = new Uint8Array(byteLength(validLen, placeHoldersLen))
+
+	// if there are placeholders, only get up to the last complete 4 chars
+	const len = placeHoldersLen > 0 ? validLen - 4 : validLen
+
+	let tmp: number
+	let curByte = 0
+
+	let i: number
+	for (i = 0; i < len; i += 4) {
+		tmp =
+			(revLookup[b64.charCodeAt(i)] << 18) |
+			(revLookup[b64.charCodeAt(i + 1)] << 12) |
+			(revLookup[b64.charCodeAt(i + 2)] << 6) |
+			revLookup[b64.charCodeAt(i + 3)]
+		arr[curByte++] = (tmp >> 16) & 0xff
+		arr[curByte++] = (tmp >> 8) & 0xff
+		arr[curByte++] = tmp & 0xff
+	}
+
+	if (placeHoldersLen === 2) {
+		tmp =
+			(revLookup[b64.charCodeAt(i)] << 2) |
+			(revLookup[b64.charCodeAt(i + 1)] >> 4)
+		arr[curByte++] = tmp & 0xff
+	}
+
+	if (placeHoldersLen === 1) {
+		tmp =
+			(revLookup[b64.charCodeAt(i)] << 10) |
+			(revLookup[b64.charCodeAt(i + 1)] << 4) |
+			(revLookup[b64.charCodeAt(i + 2)] >> 2)
+		arr[curByte++] = (tmp >> 8) & 0xff
+		arr[curByte++] = tmp & 0xff
+	}
+
+	return arr
+}
+
+const base64Description = "base64-encoded"
+const base64UrlDescription = "base64url-encoded"
+
+export const writeBase64SyntaxErrorProblem = (error: unknown): string => {
+	if (!(error instanceof SyntaxError)) throw error
+	return `must be ${base64Description} (${error})`
+}
+
+const base64Pattern =
+	/^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/
+const base64UrlPattern =
+	/^(?:[A-Za-z0-9_-]{4})*(?:[A-Za-z0-9_-]{2}(?:==|%3D%3D)?|[A-Za-z0-9_-]{3}(?:=|%3D)?)?$/
+
 export const base64 = arkModule({
-	root: regexStringNode(
-		/^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/,
-		"base64-encoded"
-	),
-	url: regexStringNode(
-		/^(?:[A-Za-z0-9_-]{4})*(?:[A-Za-z0-9_-]{2}(?:==|%3D%3D)?|[A-Za-z0-9_-]{3}(?:=|%3D)?)?$/,
-		"base64url-encoded"
-	)
+	root: regexStringNode(base64Pattern, base64Description),
+	url: regexStringNode(base64UrlPattern, base64UrlDescription),
+	parse: rootSchema({
+		in: "string",
+		declaredOut: rootSchema(Uint8Array),
+		morphs: (s: string, ctx: TraversalContext) => {
+			if (s.length === 0) return new Uint8Array(0)
+
+			try {
+				return parseB64(s)
```

Comment:
Is this just for the SyntaxError you throw in get lens?

Maybe this was just adapted from the library you mentioned, but we'd definitely like to avoid try/catch if we have control over the return types.

I'd change `getLens` to return something like a string error message in this case. Then when you call it, you'd just handle/bubble up that return manually.

Big wins in terms of perf, type-safety and transparency to encode the error in your return type in TS as opposed to throwing.